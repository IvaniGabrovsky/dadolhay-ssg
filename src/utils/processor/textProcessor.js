import fs from 'fs';

const textProcessor = input => {
  const lines = fs.readFileSync(input, 'utf8').split(/\r?\n/);

  let body = '';
  let title;

  // Each iteration of the following while should generate one block in blocks[]
  while (lines.length) {
    const blockContent = [];
    blockContent.push(lines.shift());

    // In case the file starts with \n
    if (!blockContent[0]) continue;

    // Collect all following lines that are not empty into the blockContent variable
    while (lines[0]?.match(/\w/)) {
      blockContent.push(lines.shift());
    }

    // Count number of empty lines following
    let emptyLines = 0;
    while (lines[0] === '') {
      lines.shift();
      emptyLines++;
    }

    if (!body.length && blockContent.length === 1 && emptyLines >= 2) {
      // This is a title
      body += `<h1>${blockContent.join('<br />')}</h1>\n`;
      title = blockContent;
      continue;
    }

    body += `<p>${blockContent.join('<br />')}</p>\n`;
  }

  return { body, title };
};

export default textProcessor;
