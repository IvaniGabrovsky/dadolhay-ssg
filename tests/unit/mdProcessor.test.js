import path from 'path';
import { fileURLToPath } from 'url';
import textProcessor from '../../src/utils/processor/mdProcessor';

/* eslint-disable no-underscore-dangle */

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

describe('MD processor', () => {
  test('process MD file to internal structure', () => {
    const testFileName = path.join(__dirname, '../../inputs/testMarkdown.md');
    expect(textProcessor(testFileName)).toMatchSnapshot();
  });
});
