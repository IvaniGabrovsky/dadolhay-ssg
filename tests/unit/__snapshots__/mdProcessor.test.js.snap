// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MD processor process MD file to internal structure 1`] = `
{
  "body": "<h1>Testing</h1>
<hr />
<h2>Testing</h2>
<pre><code class="language-js">  const foo = &quot;bar&quot;
</code></pre>",
}
`;
